Hummingbird documentation document generation broken

https://microsoft.github.io/hummingbird/api/hummingbird.ml.supported.html among other links display blank pages. The documentation is missing.

Update: the last PR where we deprecate `torch==1.7` causes the doc generation to fail.  

For some reason, the docs only generate with torch==1.5 (due to a conflict), so I pinned doc generation to the working eversion. But now that we mandate `torch>1.7`, it [errors out](https://github.com/microsoft/hummingbird/runs/6989934612?check_suite_focus=true) in the "Generate Documentation" phase (`AssertionError: Please install torch >1.7.0`). 

When we generate the docs with the newest torch we get

```
make sphinx-site -C website/
....
reading sources... [  3%] api/hummingbird.ml._executor                                                                                      
Extension error (sphinx.ext.linkcode):
Handler <function doctree_read at 0x7f67bd6ccee0> for event 'doctree-read' threw an exception (exception: type object 'Executor' has no attribute '_is_full_backward_hook')
```

--- a/.github/workflows/pythonapp.yml
+++ b/.github/workflows/pythonapp.yml
@@ -45,6 +45,15 @@ jobs:
     - name: Install basic dependencies
       run: |
         pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
+    # Compile (but not push) documentation only for one of the runs (Linux py3.8).
+    - name: Generate Documentation
+      if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
+      run: |
+        # We're mocking out the torch. So now we also need to sub this on out
+        find ../. -type f -exec sed -i 's/{torch.__version__}/1.11.0/g' {} +
+        find ../. -type f -exec sed -i 's/torch.__version__/"1.11.0"/g' {} +
+        make sphinx-site -C website/
+        git checkout . && pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
     - name: Run basic tests without extra
       run: pytest
     - name: Coverage on basic tests without extra
@@ -137,15 +146,6 @@ jobs:
       with:
         file: ./coverage.xml
         flags: unittests
-    # Compile and push documentation only for one of the runs (Linux py3.8).
-    - name: Generate Documentation
-      if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
-      run: |
-        # We're mocking out the torch. So now we also need to sub this on out
-        find ../. -type f -exec sed -i 's/{torch.__version__}/1.11.0/g' {} +
-        find ../. -type f -exec sed -i 's/torch.__version__/"1.11.0"/g' {} +
-        pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
-        make sphinx-site -C website/
     - name: Deploy to GitHub pages 🚀
       if: ${{ github.ref == 'refs/heads/main' && matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
       uses: JamesIves/github-pages-deploy-action@4.0.0

--- a/.github/workflows/pythonapp.yml
+++ b/.github/workflows/pythonapp.yml
@@ -44,7 +44,7 @@ jobs:
       run: pip install torch==1.11.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
     - name: Install basic dependencies
       run: |
-        pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
+        pip install -e .[tests] -f https://download.pytorch.org/whl/torch_stable.html
     - name: Run basic tests without extra
       run: pytest
     - name: Coverage on basic tests without extra
@@ -141,6 +141,10 @@ jobs:
     - name: Generate Documentation
       if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
       run: |
+        # We're mocking out the torch. So now we also need to sub this on out
+        find ../. -type f -exec sed -i 's/{torch.__version__}/1.11.0/g' {} +
+        find ../. -type f -exec sed -i 's/torch.__version__/"1.11.0"/g' {} +
+        pip install -e .[tests] -f https://download.pytorch.org/whl/torch_stable.html
         make sphinx-site -C website/
     - name: Deploy to GitHub pages 🚀
       if: ${{ github.ref == 'refs/heads/main' && matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}

--- a/.github/workflows/pythonapp.yml
+++ b/.github/workflows/pythonapp.yml
@@ -44,7 +44,7 @@ jobs:
       run: pip install torch==1.11.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
     - name: Install basic dependencies
       run: |
-        pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
+        pip install -e .[tests] -f https://download.pytorch.org/whl/torch_stable.html
     # Compile (but not push) documentation only for one of the runs (Linux py3.8).
     - name: Generate Documentation
       if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
@@ -53,7 +53,8 @@ jobs:
         find ../. -type f -exec sed -i 's/{torch.__version__}/1.11.0/g' {} +
         find ../. -type f -exec sed -i 's/torch.__version__/"1.11.0"/g' {} +
         make sphinx-site -C website/
-        git checkout . && pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
+        # now refresh the code
+        git checkout . && pip install -e .[tests] -f https://download.pytorch.org/whl/torch_stable.html
     - name: Run basic tests without extra
       run: pytest
     - name: Coverage on basic tests without extra

--- a/.github/workflows/pythonapp.yml
+++ b/.github/workflows/pythonapp.yml
@@ -44,7 +44,7 @@ jobs:
       run: pip install torch==1.11.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
     - name: Install basic dependencies
       run: |
-        pip install -e .[tests] -f https://download.pytorch.org/whl/torch_stable.html
+        pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
     - name: Run basic tests without extra
       run: pytest
     - name: Coverage on basic tests without extra
@@ -144,7 +144,7 @@ jobs:
         # We're mocking out the torch. So now we also need to sub this on out
         find ../. -type f -exec sed -i 's/{torch.__version__}/1.11.0/g' {} +
         find ../. -type f -exec sed -i 's/torch.__version__/"1.11.0"/g' {} +
-        pip install -e .[tests] -f https://download.pytorch.org/whl/torch_stable.html
+        pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
         make sphinx-site -C website/
     - name: Deploy to GitHub pages 🚀
       if: ${{ github.ref == 'refs/heads/main' && matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}

--- a/.github/workflows/pythonapp.yml
+++ b/.github/workflows/pythonapp.yml
@@ -18,8 +18,8 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [ubuntu-18.04]
-        python-version: [3.8]
+        os: [ubuntu-18.04, macos-11, windows-2019]
+        python-version: [3.8, 3.9]
 
     steps:
     - uses: actions/checkout@v2
@@ -27,12 +27,116 @@ jobs:
       uses: actions/setup-python@v2
       with:
         python-version: ${{ matrix.python-version }}
+    - name: Lint with flake8
+      run: |
+        python -m pip install --upgrade pip
+        pip install flake8
+        # stop the build if there are Python syntax errors or undefined names
+        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+        # The GitHub editor is 127 chars wide
+        flake8 . --count  --max-complexity=10 --max-line-length=127 --statistics
+    # PyTorch for Mac has different pip syntax wrt Win and Linux.
     - name: Install pytorch 1.11.0 if mac
       if: ${{  matrix.os == 'macos-11' }}
       run: pip install torch==1.11.0
     - name: Install pytorch 1.11.0+cpu if not mac
       if:  ${{  matrix.os != 'macos-11' }}
       run: pip install torch==1.11.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
+    - name: Install basic dependencies
+      run: |
+        pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
+    - name: Run basic tests without extra
+      run: pytest
+    - name: Coverage on basic tests without extra
+      run: coverage run -a -m pytest tests/test_no_extra_install.py
+    - name: If mac, install libomp to facilitate lgbm install
+      if: matrix.os == 'macos-11'
+      run: |
+        wget https://raw.githubusercontent.com/Homebrew/homebrew-core/fb8323f2b170bd4ae97e1bac9bf3e2983af3fdb0/Formula/libomp.rb
+        brew unlink libomp
+        brew install ./libomp.rb
+        export CC=/usr/bin/clang
+        export CXX=/usr/bin/clang++
+        export CPPFLAGS="$CPPFLAGS -Xpreprocessor -fopenmp"
+        export CFLAGS="$CFLAGS -I/usr/local/opt/libomp/include"
+        export CXXFLAGS="$CXXFLAGS -I/usr/local/opt/libomp/include"
+        export LDFLAGS="$LDFLAGS -Wl,-rpath,/usr/local/opt/libomp/lib -L/usr/local/opt/libomp/lib -lomp"
+    - name: Install extra dependencies
+      run: |
+        pip install .[extra,onnx,sparkml]
+        pip install pandas
+    - name: pip list
+      run: pip list
+    - uses: actions/cache@v1
+      # TVM takes forever, we try to cache it.
+      if: ${{ matrix.os != 'windows-2019' }}
+      id: cache
+      env:
+        CACHE_NUMBER: 3
+      with:
+        path: ../../../tvm
+        key: ${{ runner.os }}-${{ env.CACHE_NUMBER }}-${{ matrix.python-version }}-tvm-0.8
+    # Getting TVM requires: 1) fetching TVM from github, 2) get LLVM, 3) cmake, 4) make, 5) install python dependecy.
+    # 1 to 4 will be retrieved from the cache.
+    # The pipeline only works for Unix systems. For windows we will have to compile LLVM from source which is a no go.
+    - name: Fetch and prepare TVM for compilation
+      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os != 'windows-2019' }}
+      run: |
+        cd ~/
+        git clone https://github.com/apache/tvm.git
+        cd tvm
+        git checkout tags/v0.8.0
+        git submodule update --recursive --init
+        cmake -E make_directory build
+    - name: Get LLVM on Linux
+      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os == 'ubuntu-18.04' }}
+      working-directory: ../../../tvm
+      run: |
+        wget https://github.com/llvm/llvm-project/releases/download/llvmorg-10.0.0/clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz
+        tar -xf clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz && mv clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04 llvm
+    - name: Get LLVM on Mac
+      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os == 'macos-11' }}
+      working-directory: ../../../tvm
+      run: |
+        wget https://github.com/llvm/llvm-project/releases/download/llvmorg-10.0.0/clang+llvm-10.0.0-x86_64-apple-darwin.tar.xz
+        tar -xf clang+llvm-10.0.0-x86_64-apple-darwin.tar.xz && mv clang+llvm-10.0.0-x86_64-apple-darwin llvm
+    - name: CMake TVM
+      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os != 'windows-2019' }}
+      working-directory: ../../../tvm/build
+      run: >-
+        cmake
+        "-DUSE_RPC=ON"
+        "-DUSE_GRAPH_RUNTIME=ON"
+        "-DUSE_LLVM=../llvm/bin/llvm-config"
+        ..
+    - name: Build TVM
+      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os != 'windows-2019' }}
+      working-directory: ../../../tvm/build
+      run: |
+        make -j3
+    - name: Install python TVM
+      if: ${{  matrix.os != 'windows-2019' }}
+      working-directory: ../../../tvm/python
+      run: |
+        python setup.py install
+    - name: pip list
+      run: pip list
+    # We don't run pytest for Linux py3.8 since we do coverage for that case.
+    - name: Test with pytest
+      if: ${{ matrix.python-version != '3.8' || matrix.os != 'ubuntu-18.04' }}
+      run: pytest
+    # Run and push coverage only for Linux py3.8
+    - name: Coverage 3.8 Linux
+      if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
+      run: |
+        coverage run -a -m pytest tests
+        coverage xml
+    - name: Upload coverage to Codecov
+      if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
+      uses: codecov/codecov-action@v1
+      with:
+        file: ./coverage.xml
+        flags: unittests
     # Compile and push documentation only for one of the runs (Linux py3.8).
     - name: Generate Documentation
       if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}

--- a/.github/workflows/pythonapp.yml
+++ b/.github/workflows/pythonapp.yml
@@ -18,8 +18,8 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [ubuntu-18.04, macos-11, windows-2019]
-        python-version: [3.8, 3.9]
+        os: [ubuntu-18.04]
+        python-version: [3.8]
 
     steps:
     - uses: actions/checkout@v2
@@ -27,116 +27,12 @@ jobs:
       uses: actions/setup-python@v2
       with:
         python-version: ${{ matrix.python-version }}
-    - name: Lint with flake8
-      run: |
-        python -m pip install --upgrade pip
-        pip install flake8
-        # stop the build if there are Python syntax errors or undefined names
-        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
-        # The GitHub editor is 127 chars wide
-        flake8 . --count  --max-complexity=10 --max-line-length=127 --statistics
-    # PyTorch for Mac has different pip syntax wrt Win and Linux.
     - name: Install pytorch 1.11.0 if mac
       if: ${{  matrix.os == 'macos-11' }}
       run: pip install torch==1.11.0
     - name: Install pytorch 1.11.0+cpu if not mac
       if:  ${{  matrix.os != 'macos-11' }}
       run: pip install torch==1.11.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
-    - name: Install basic dependencies
-      run: |
-        pip install .[tests] -f https://download.pytorch.org/whl/torch_stable.html
-    - name: Run basic tests without extra
-      run: pytest
-    - name: Coverage on basic tests without extra
-      run: coverage run -a -m pytest tests/test_no_extra_install.py
-    - name: If mac, install libomp to facilitate lgbm install
-      if: matrix.os == 'macos-11'
-      run: |
-        wget https://raw.githubusercontent.com/Homebrew/homebrew-core/fb8323f2b170bd4ae97e1bac9bf3e2983af3fdb0/Formula/libomp.rb
-        brew unlink libomp
-        brew install ./libomp.rb
-        export CC=/usr/bin/clang
-        export CXX=/usr/bin/clang++
-        export CPPFLAGS="$CPPFLAGS -Xpreprocessor -fopenmp"
-        export CFLAGS="$CFLAGS -I/usr/local/opt/libomp/include"
-        export CXXFLAGS="$CXXFLAGS -I/usr/local/opt/libomp/include"
-        export LDFLAGS="$LDFLAGS -Wl,-rpath,/usr/local/opt/libomp/lib -L/usr/local/opt/libomp/lib -lomp"
-    - name: Install extra dependencies
-      run: |
-        pip install .[extra,onnx,sparkml]
-        pip install pandas
-    - name: pip list
-      run: pip list
-    - uses: actions/cache@v1
-      # TVM takes forever, we try to cache it.
-      if: ${{ matrix.os != 'windows-2019' }}
-      id: cache
-      env:
-        CACHE_NUMBER: 3
-      with:
-        path: ../../../tvm
-        key: ${{ runner.os }}-${{ env.CACHE_NUMBER }}-${{ matrix.python-version }}-tvm-0.8
-    # Getting TVM requires: 1) fetching TVM from github, 2) get LLVM, 3) cmake, 4) make, 5) install python dependecy.
-    # 1 to 4 will be retrieved from the cache.
-    # The pipeline only works for Unix systems. For windows we will have to compile LLVM from source which is a no go.
-    - name: Fetch and prepare TVM for compilation
-      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os != 'windows-2019' }}
-      run: |
-        cd ~/
-        git clone https://github.com/apache/tvm.git
-        cd tvm
-        git checkout tags/v0.8.0
-        git submodule update --recursive --init
-        cmake -E make_directory build
-    - name: Get LLVM on Linux
-      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os == 'ubuntu-18.04' }}
-      working-directory: ../../../tvm
-      run: |
-        wget https://github.com/llvm/llvm-project/releases/download/llvmorg-10.0.0/clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz
-        tar -xf clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz && mv clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04 llvm
-    - name: Get LLVM on Mac
-      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os == 'macos-11' }}
-      working-directory: ../../../tvm
-      run: |
-        wget https://github.com/llvm/llvm-project/releases/download/llvmorg-10.0.0/clang+llvm-10.0.0-x86_64-apple-darwin.tar.xz
-        tar -xf clang+llvm-10.0.0-x86_64-apple-darwin.tar.xz && mv clang+llvm-10.0.0-x86_64-apple-darwin llvm
-    - name: CMake TVM
-      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os != 'windows-2019' }}
-      working-directory: ../../../tvm/build
-      run: >-
-        cmake
-        "-DUSE_RPC=ON"
-        "-DUSE_GRAPH_RUNTIME=ON"
-        "-DUSE_LLVM=../llvm/bin/llvm-config"
-        ..
-    - name: Build TVM
-      if: ${{ steps.cache.outputs.cache-hit != 'true' && matrix.os != 'windows-2019' }}
-      working-directory: ../../../tvm/build
-      run: |
-        make -j3
-    - name: Install python TVM
-      if: ${{  matrix.os != 'windows-2019' }}
-      working-directory: ../../../tvm/python
-      run: |
-        python setup.py install
-    - name: pip list
-      run: pip list
-    # We don't run pytest for Linux py3.8 since we do coverage for that case.
-    - name: Test with pytest
-      if: ${{ matrix.python-version != '3.8' || matrix.os != 'ubuntu-18.04' }}
-      run: pytest
-    # Run and push coverage only for Linux py3.8
-    - name: Coverage 3.8 Linux
-      if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
-      run: |
-        coverage run -a -m pytest tests
-        coverage xml
-    - name: Upload coverage to Codecov
-      if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}
-      uses: codecov/codecov-action@v1
-      with:
-        file: ./coverage.xml
-        flags: unittests
     # Compile and push documentation only for one of the runs (Linux py3.8).
     - name: Generate Documentation
       if: ${{ matrix.python-version == '3.8' && matrix.os == 'ubuntu-18.04' }}

--- a/website/Makefile
+++ b/website/Makefile
@@ -23,6 +23,9 @@ clean-sphinx-site:
 .PHONY:
 sphinx-site: #clean-sphinx-site
 	pip install sphinx sphinx_rtd_theme psutil
+	# We're mocking out the torch. So now we also need to sub this on out
+	find ../. -type f -exec sed -i 's/{torch.__version__}/1.11.0/g' {} +
+	find ../. -type f -exec sed -i 's/torch.__version__/"1.11.0"/g' {} +
 
 	sphinx-apidoc --private -o sphinx/api -t sphinx/_templates ../hummingbird  -d 1 -f -e
 	$(MAKE) -C sphinx

--- a/website/Makefile
+++ b/website/Makefile
@@ -23,9 +23,7 @@ clean-sphinx-site:
 .PHONY:
 sphinx-site: #clean-sphinx-site
 	pip install sphinx sphinx_rtd_theme psutil
-	# TODO: There is a bug with torch 1.6 causing a "has no attribute 'training')" error.  Using 1.5 for now.
-	# Update: there are even more bugs with newer versions. I'm not quite sure what to do.
-	pip install torch==1.5.1+cpu -f https://download.pytorch.org/whl/torch_stable.html
+
 	sphinx-apidoc --private -o sphinx/api -t sphinx/_templates ../hummingbird  -d 1 -f -e
 	$(MAKE) -C sphinx
 	$(MAKE) copy-python-apis

--- a/website/Makefile
+++ b/website/Makefile
@@ -23,9 +23,6 @@ clean-sphinx-site:
 .PHONY:
 sphinx-site: #clean-sphinx-site
 	pip install sphinx sphinx_rtd_theme psutil
-	# We're mocking out the torch. So now we also need to sub this on out
-	find ../. -type f -exec sed -i 's/{torch.__version__}/1.11.0/g' {} +
-	find ../. -type f -exec sed -i 's/torch.__version__/"1.11.0"/g' {} +
 
 	sphinx-apidoc --private -o sphinx/api -t sphinx/_templates ../hummingbird  -d 1 -f -e
 	$(MAKE) -C sphinx

--- a/website/sphinx/conf.py
+++ b/website/sphinx/conf.py
@@ -47,6 +47,7 @@ else:
     extensions.append("sphinx.ext.imgmath")
 
 autodoc_default_flags = ["members", "inherited-members", "private-members"]
+autodoc_mock_imports = ["torch"]
 
 # Add any paths that contain templates here, relative to this directory.
 templates_path = ["_templates"]

